### PR TITLE
Improve sync flow from Accounts UI

### LINKED ISSUE #94766: Improve the settings sync flows with authentication
Improve the settings sync flows with authentication

## Sign in flow

Signing in showing too many notifications and not consistent. Sign in from Accounts UI and I see following dialogs 

a. For the first time ever

image

image

b. Afterwards

image

image

## Sign out flow

Signing out is showing following notification

image

There should be a sign in button instead of Turn on. Because user did not turn off sync here. It shall be same as the action button that is shown in the gear icon

image

Above notification is also shown even when settings sync is turned off

Also do we need successfully sign out notification here?

## Error flow

Errors while signing in

Following notifications are shown which are not helpful

image

## Cross Window Flow

Sign in/Sign out from one window is prompting turn on/turn off sync notification on another window

https://github.com/microsoft/vscode/issues/94176

- Open two windows with sync enabled 
- Sign out from one window and you get a turn off notification in that window 
- But it is not expected to get the same notification in another window
- Now sign in from one window and you get following notification in another window

image

## Others

- Decouple auth and settings sync - https://github.com/microsoft/vscode/issues/94050

### PR DESCRIPTION
**Pull request by sandy081**
Improve the sync flow from Accounts UI

Remove Sign in actions in favour of having turn on sync action also in Accounts dropdown
Sign out shows a confirmation dialog that sync will be stopped

Less dialogs and notifications.

Fixes #94766

### CODE REVIEW

File: src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		let status = AccountStatus.Unavailable;
+		let allAccounts: Map<string, IUserDataSyncAccount> = new Map<string, IUserDataSyncAccount>();
+
+		if (this.accountProviderId) {
+
+			let currentAccount: IUserDataSyncAccount | null = null;
+			let currentSession: AuthenticationSession | undefined = undefined;
+
+			if (this.currentSessionId) {
+				const sessions = await this.authenticationService.getSessions(this.accountProviderId) || [];
```
Comment (RMacfarlane) [inline src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts:110]: I think allAccounts should still be set even if there isn't a current session. For example, if a user signs into an MSA with another extension before setting sync, we would want that account to be an available option

**Review by RMacfarlane — CHANGES_REQUESTED** [review]
Looks good overall!

I think with moving the commands into the account menu, it makes less sense to duplicate them in the gear menu. If we still want to show a badge indicator when settings sync is on but there is no active account, I think it  would be clearer to show it on the accounts menu instead.

File: src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+					if (selected.account) {
+						await this.switch(selected.account.sessionId, selected.account.accountName);
+					} else {
+						await this.login();
+					}
+					quickPick.hide();
+					c();
+				}
+			}));
+			disposables.add(quickPick.onDidHide(() => disposables.dispose()));
```
[the remark this answers is not in the document]
Comment (sandy081) [reply src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts]: Good catch. 👍

File: src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts (hunk 1 above)
Comment (sandy081) [reply src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts]: Good catch. 👍

File: src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+export interface IUserDataSyncAccount {
+	readonly providerId: string;
+	readonly sessionId: string;
+	readonly accountName: string;
+}
+
+export const enum AccountStatus {
+	Uninitialized = 'uninitialized',
+	Unavailable = 'unavailable',
+	Inactive = 'inactive',
```
[the remark this answers is not in the document]
Comment (sandy081) [reply src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts:38]: Do you mean to replace `Active`, `Inactive` and `Unavailable` with `AccountSet`, `NoAccountSet` and `AccountUnavailable` respectively?

**Comment by sandy081** [comment]
IMO showing sync actions at the locations where they belong contextually, is good and also helpful in terms of discoverability. Following are the preferences related locations

- Gear icon, 
- file > preferences menu, 
- Settings Editor, 
- Keybindings Editor 
- Extensions. 

So I think it makes sense to show `Turn on Preferences Sync` action at these places.

I agree that `Sign in` action goes well under Accounts UI. Will look into it.

File: src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandy081) [reply src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts]: Discussed offline and agreed upon having following 3 states

- `Uninitialized`
- `Unavailable`
- `Available`

**Review by RMacfarlane — APPROVED** [review]
Looks good!

**Pull request merged**